bin/helpers/openstack.sh

```diff
@@ -0,0 +1,357 @@
+    exit 1
+  fi
+}
+
+function download_image() {
+  local download_filename=$1
+  local download_url=$2
+  local download_filepath="/tmp/${download_filename}"
+
+  echo "Downloading image to local /tmp/"
+  echo -e "\n"
+  curl "${download_url}/${download_filename}" \
+    -o "${download_filepath}" \
+    --connect-timeout 30 \
+    --max-time 1800
+
+  echo "Verifying the md5 checksum of the downloaded image"
+  echo -e "\n"
+  verify_downloaded_image "${download_filename}" "${download_url}"
+}
+
+function maybe_download_image(){
+  local image_name=$1
+  local disc_format=$2
+  local download_url=$3
+  local download_filename="${image_name}.qcow2"
+  local download_filepath="/tmp/${image_name}.qcow2"
+  local is_image_downloaded_already
+  is_image_downloaded_already=$(is_image_downloaded_already "${download_filepath}")
+
+  if [[ "${is_image_downloaded_already}" == "false" ]]; then
+    download_image "${download_filename}" "${download_url}"
+  else
+    echo "${download_filename} is already downloaded"
+    echo "Verifying the downloaded image checksum"
+    echo -e "\n"
+
+    verify_downloaded_image "${download_filename}" "${download_url}"
+  fi
+}
+
+function maybe_convert_image(){
+  local image_name=$1
+  local disc_format=$2
+  local download_url=$3
+  local download_filename="${image_name}.qcow2"
+  local download_filepath="/tmp/${image_name}.qcow2"
+  local upload_name="${image_name}"
+  local upload_filepath="/tmp/${image_name}.${disc_format}"
+  local qemu_installed
+  local is_image_downloaded_already
+  is_image_downloaded_already=$(is_image_downloaded_already "${download_filepath}")
+  qemu_installed=$(which qemu-img)
+
+
+  if [[ ! -e ${upload_filepath} ]]; then
+    if [[ "${qemu_installed}" == "" ]]; then
+      echo >&2 "qemu-utils package is missing - image cannot be converted"
+      echo >&2 "Please ensure that qemu-img is available."
+      exit 1
+    else
+      if [[ "${is_image_downloaded_already}" == "false" ]]; then
+        echo >&2 "No file found at ${download_filepath}"
+        echo >&2 "Please ensure image has been downloaded"
+        exit 1
+      fi
+
+      echo "${download_filepath} has not yet been converted to the ${disc_format} format"
+      echo "Starting the image conversion from qcow2 to ${disc_format}..."
+      echo -e "\n"
+
+      qemu-img convert -q -f qcow2 -O "${disc_format}" "${download_filepath}" "${upload_filepath}"
+    fi
+  else
+    echo "${upload_filepath} already exists and so does not require conversion, moving onto upload step"
+  fi
+}
+
+function maybe_upload_image() {
+  local image_name=$1
+  local disc_format=$2
+  local image_url=$3
+  local is_image_uploaded
+  is_image_uploaded=$(is_image_already_uploaded "${image_name}" "${disc_format}")
+
+  if [[ "${is_image_uploaded}" == "false" ]]; then
+    maybe_convert_image "${image_name}" "${disc_format}" "${image_url}"
+    upload_image "${image_name}" "${disc_format}" "${image_url}"
+  else
```

Comment:
only call "post_image_upload_checks" from function "upload_image" (don't call if image is uploaded already previously - checksum could be different since same version image can have different checksums due to security updates)
(remove else block)

Reply:
- this function is removed bu the logic for it has moved to `image-create-openstack.sh`. 

- I've moved `post_image_upload_checks` into upload image.

- If we remove the else block then this line 
`echo "${image_name}.${glance_disc_format} is already present - no need to upload"` also gets removed - perhaps it's better to keep this as it signals to the user that the entire process for uploading the openstack image will be bypassed ?